lib/batch.js

```diff
@@ -4,51 +4,90 @@
-    selfObj.batchHash = {};
+
     return selfObj;
   }
 
-  selfObj.batch = function(batchKey, data, processTime, jobName) {
-    if ( this.batchHash[batchKey] !== undefined ) {
-      // The key / job exists
-      var job = this.batchHash[ batchKey ];
-      job.attrs.data.jobData.push( data );
-      this.agendaModule.saveJob(job);
-    } else {
-      // The key / job has not been spawned yet.
-      var batchData = { key: batchKey, jobData: [] };
-      batchData.jobData.push( data );
-      this.batchHash[ batchKey ] = this.agendaModule.schedule(processTime, jobName, batchData );
-    }
+  var spawnNewJob = function(batchKey, data, processTime, jobName) {
+    var batchData = { key: batchKey, jobData: [] };
+    batchData.jobData.push( data );
+    return selfObj.agendaModule.schedule(processTime, jobName, batchData );
+  }
+
+  var updateExistingJob = function( job, data ) {
+    job.attrs.data.jobData.push( data );
+    selfObj.agendaModule.saveJob(job);
   }
 
-  selfObj.define = function(jobName, options, processor) {
-    if (!processor) {
-      processor = options;
-      options = {};
+  selfObj.batch = function(batchKey, data, processTime, jobName, cb) {
+
+    if ( cb == undefined ) {
+      cb = function(err, job) {};
     }
 
-    selfObj.agendaModule.define(jobName, options, processor);
+    // REFACTOR
+    selfObj.agendaModule.jobs( { 'data.key': batchKey }, function(err, jobs) {
+      if ( err !== null ) {
+        cb(err, null);
+        return;
+      }
+
+      if ( jobs.length == 0 ) {
+        // No Jobs we're found... better make one up!
+        newJob = spawnNewJob(batchKey, data, processTime, jobName);
+        cb(null, newJob);
+
+        return;
+      }
+
+      if ( jobs.length > 1 ) {
+        // There is more than one job with this key in the system. wtf?
+        // Let's append the data to the last job that is not running.
+        var job = null;
+
+        for ( var i = (jobs.length - 1); i == 0; i--) {
+          var proxJob = jobs[i];
+          if ( proxJob.isRunning() !== false ) {
+            job = proxJob;
+            updateExistingJob( job, batchData );
+          }
+        }
+
+        if ( job === null ) {
+          // No Non Running Job was found. Double wtf? Oh Well, let's spawn a new one.
+          job = spawnNewJob(batchKey, data, processTime, jobName);
+        }
+
+        cb(null, job);
+        return;
+      }
+      var job = jobs[0];
+      if ( job.isRunning() === false ) {
+        updateExistingJob( job, data );
+      } else {
+        // Job be running, time to spawn a new one!
+        job = spawnNewJob(batchKey, data, processTime, jobName);
+      }
+
+      cb( null, job );
+
+    });
   }
 
   selfObj.cleanBatch = function( job ) {
-    if( selfObj.batchHash[ job.attrs.data.key ] !== undefined ) {
-      delete selfObj.batchHash[ job.attrs.data.key ];
-    }
+    job.remove(function(err, count) {});
```

Comment:
This was to ensure that the data was nuked. But, the new paradigm of job checking makes it un-necessary.